app/models/customer/merge.rb

```diff
@@ -127,14 +142,40 @@ def self.save_and_update_foreign_keys!(c0,c1)
         old_created_by_admin = c0.created_by_admin
         c0.created_by_admin = true
         c0.save!
-        c0.update_attribute(:created_by_admin, false) if !old_created_by_admin
+        unless old_created_by_admin
+          c0.update_attribute(:created_by_admin, false)
+        end
+
+        # This is the tricky part, we should distinguish whether the user is a user before
+        # we applied 'identity' or after.
+        # Password needs to be changed
         if pass
-          Customer.connection.execute("UPDATE customers SET crypted_password='#{pass}',salt='#{salt}' WHERE id=#{c0.id}")
+          if keeping_bcrypted
+            # The fresher password is using the old version encrption
+            if salt
+              c0.crypted_password = pass
+              c0.salt = salt
+              c0.save
+              # Customer.connection.execute("UPDATE customers SET crypted_password='#{pass}',salt='#{salt}' WHERE id=#{c0.id}")
+            else
+              Authorization.create(provider: 'identity', uid: email, customer: c0, password_digest: pass)
+              # Customer's password and salt is protected, cannot change, just Keep it
+              # Customer.find(c0.id).update(crypted_password: "", salt: "")
+            end
+          else
+            unless salt
+              identity = c0.identity
+              identity.uid = email
+              identity.password_digest = pass
+              identity.save(:validate => false)
+              # Authorization.connection.execute("UPDATE authorizations SET uid='#{email}',password_digest='#{pass}' WHERE customer_id=#{c0.id} AND provider='identity'")
```

Comment:
The Authorization.connection.execute is based on a Customer.connection.execute line from the original code. I replaced both these statements with a long-form version because code climate was complaining about possible sql injection vulnerabilities. I left the comment so it would be easier to understand but I guess you can just see it in the diff. I'm deleting the comment.